src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16TextEncodingLE.cs

```diff
@@ -7,113 +7,145 @@ namespace System.Text.Utf16
-            bytesWritten = 0;
+            int consumed;
+            var utf16 = data.Cast<byte, char>();
+            var result = Utf16LittleEndianEncoder.TryDecode(utf16, utf32, out consumed, out charactersWritten);
 
-            for (int i = 0; i < outputLength; i++)
-            {
-                UnicodeCodePoint decodedCodePoint = decoded[i];
+            bytesConsumed = consumed * sizeof(char);
+            return result;
+        }
 
-                if (avaliableBytes - bytesWritten < 2)
-                {
-                    decodedCodePoint = new UnicodeCodePoint();
-                    bytesWritten = 0;
-                    return false;
-                }
+        #endregion Decoding implementation
 
-                uint answer = (uint)(encoded[1 + bytesWritten] << 8 | encoded[bytesWritten]);
-                decodedCodePoint = new UnicodeCodePoint(answer);
-                bytesWrittenForCodePoint = 2;
+        #region Encoding implementation
 
-                if (avaliableBytes - bytesWritten >= 4)
-                {
-                    uint highBytes = answer;
-                    uint lowBytes = (uint)(encoded[3 + bytesWritten] << 8 | encoded[2 + bytesWritten]);
-
-                    if (highBytes >= UnicodeConstants.Utf16HighSurrogateFirstCodePoint
-                            && highBytes <= UnicodeConstants.Utf16HighSurrogateLastCodePoint
-                            && lowBytes >= UnicodeConstants.Utf16LowSurrogateFirstCodePoint
-                            && lowBytes <= UnicodeConstants.Utf16LowSurrogateLastCodePoint)
-                    {
-                        answer = (((highBytes - UnicodeConstants.Utf16HighSurrogateFirstCodePoint) << 10)
-                            | (lowBytes - UnicodeConstants.Utf16LowSurrogateFirstCodePoint)) + 0x10000;
-
-                        decodedCodePoint = new UnicodeCodePoint(answer);
-                        bytesWrittenForCodePoint = 4;
-                    }
-                }
+        public override bool TryEncode(ReadOnlySpan<byte> utf8, Span<byte> data, out int bytesConsumed, out int bytesWritten)
+        {
+            int charactersWritten;
+            var utf16 = data.Cast<byte, char>();
+            var result = Utf8Encoder.TryDecode(utf8, utf16, out bytesConsumed, out charactersWritten);
+
+            bytesWritten = charactersWritten * sizeof(char);
+            return result;
+        }
+
+        public override bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> data, out int charactersConsumed, out int bytesWritten)
+        {
+            // TODO: Other methods validate that the input stream contains valid sequences as they are consumed.
+            //       Currently, this is a copy operation with no validation. What is the right thing here?
 
-                decoded[i] = decodedCodePoint;
-                bytesWritten += bytesWrittenForCodePoint;
+            var charBuffer = data.Cast<byte, char>();
+
+            if (utf16.Length > charBuffer.Length)
```

Comment:
That would mean that we can't optimize here with a simple buffer copy. If we are doing UTF-16 to UTF-16 (i.e. nothing really), then assuming that the input is correct means a buffer copy. Otherwise, we have to do a full decode / encode to validate the data before doing the copy.